ibis/expr/types/tvf.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from public import public
+
+import ibis.common.exceptions as com
+import ibis.expr.analysis as an
+import ibis.expr.operations as ops
+import ibis.expr.types as ir
+from ibis.common.deferred import Deferred
+from ibis.selectors import Selector
+
+if TYPE_CHECKING:
+    from ibis.expr.types import Table
+
+
+def _get_window_by_key(table, value):
+    if isinstance(value, str):
+        return table[value]
+    elif isinstance(value, Deferred):
+        return value.resolve(table)
+    elif isinstance(value, Selector):
+        return value.expand(table)
+    elif isinstance(value, ir.Expr):
+        return an.sub_immediate_parents(value.op(), table.op()).to_expr()
+    else:
+        return value
+
+
+@public
+class WindowedTable:
+    """An intermediate table expression to hold windowing information."""
+
+    def __init__(self, table: ir.Table, time_col: ir.Value):
+        self.table = table
+        self.time_col = _get_window_by_key(table, time_col)
+
+        if self.time_col is None:
+            raise com.IbisInputError(
+                "Window aggregations require `time_col` as an argument"
+            )
+
+    def tumble(
+        self,
+        window_size: ir.IntervalScalar,
+        offset: ir.IntervalScalar | None = None,
+    ) -> Table:
+        """Compute a tumble table valued function.
+
+        Parameters
+        ----------
+        window_size
+            Width of the tumbling windows.
+        offset
+            An optional parameter to specify the offset which window start should be shifted by.
+
+        Returns
+        -------
+        Table
+            Table expression.
```

Comment:
These docstrings could use some improvement (example, better description of what each method and parameter does), but this can be done in a follow-up if you want. The examples in particular would be easier to write once we add support for a batch backend like duckdb.